xml_converter/src/packaging_protobin.cpp

```diff
@@ -284,3 +291,47 @@ void write_protobuf_file_per_map_id(
             &state);
     }
 }
+void _category_filter_by_depth(
+    const map<string, Category>* marker_categories,
+    int target_depth,
+    map<string, StringHierarchy>* category_hierarchy,
+    vector<string> parent_name,
+    int current_depth = 0) {
+    for (auto it = marker_categories->begin(); it != marker_categories->end(); it++) {
+        vector<string> child_name = parent_name;
+        child_name.push_back(it->first);
+        if (target_depth == current_depth) {
+            StringHierarchy category_filter;
+            category_filter.add_path(child_name, true);
+            (*category_hierarchy)[join(child_name, "_")] = category_filter;
+        }
+
+        _category_filter_by_depth(&(it->second.children), target_depth, category_hierarchy, child_name, current_depth + 1);
+    }
+}
+
+// Write protobuf per category
+void write_protobuf_file_per_category(
+    const string& marker_pack_root_directory,
+    const OptionalInt& split_by_category_depth,
+    const map<string, Category>* marker_categories,
+    const vector<Parseable*>* parsed_pois) {
+    ProtoWriterState state;
+    state.marker_pack_root_directory = marker_pack_root_directory;
+
+    map<string, StringHierarchy> category_hierarchies;
+    _category_filter_by_depth(marker_categories, split_by_category_depth.get_value(), &category_hierarchies, {});
```

Comment:
I feel like this line should read:
```cpp
map<string, StringHierarchy> category_hierarchies = category_filter_by_depth(marker_categories, split_by_category_depth.get_value());
```

It is generally more readable to return a value than ask for it to be passed in as a pointer, though passing it in as a pointer is done often in c, and unavoidable when writing functions intended to mutate a class outside of the class methods.

In this case I recommend following the pattern setup where `category_filter_by_depth` handles all the external setup and internally calls `_category_filter_by_depth`.

Reply:
I like that pattern. I'll make the change.